axlearn/cloud/gcp/jobset_utils.py

```diff
@@ -898,18 +986,258 @@ def _build_a3_init_container(self) -> Nested[Any]:
+            "-c",
+            'set -x; /tcpgpudmarxd/build/app/tcpgpudmarxd --gpu_nic_preset a3vm  \
+                --gpu_shmem_type fd --uds_path /run/tcpx \
+                --setup_param "--verbose 128 2 0" & \n\
+            while [ ! -f /run/tcpx/terminated ]; do sleep 10; done;',
+        ]
+
+        return dict(
+            name="tcpx-daemon",
+            image="us-docker.pkg.dev/gce-ai-infra/gpudirect-tcpx/tcpgpudmarxd-dev:latest",
+            securityContext={"privileged": True},
+            command=command,
+            env=[{"name": "LD_LIBRARY_PATH", "value": "/usr/local/nvidia/lib64"}],
+            volumeMounts=volume_mounts,
+        )
+
+
+    def _build_init_containers(self) -> Nested[Any]:
+        return [self._build_a3_high_tcpx_init_container(), self._build_a3_high_sidecar_container()]
+
+class A3MegaReplicatedJob(GPUReplicatedJob):
+    """Builds a replicated job spec for an a3-mega GPU job, to be used with JobSet API."""
+
+    Config = GPUReplicatedJob.Config
+
+    def _build_a3_mega_tcpx_init_container(self) -> Nested[Any]:
+        """Builds a config for a single container."""
+        volume_mounts = [
+            {
+                "name": "tcpx-nccl-plugin-volume",
+                "mountPath": "/var/lib/tcpx",
+            },
+        ]
+        command = command = [
+            "bash",
+            "-c",
+            'set -ex; chmod 755 /scripts/container_entry.sh; \n\
+             /scripts/container_entry.sh install; \n\
+             mkdir -p /usr/lib/tcpx/lib64; \n\
+             cp -r /var/lib/tcpxo/lib64/. /usr/lib/tcpx/lib64; \n\
+             echo "installation finishes";',
+        ]
+        return dict(
+            name="tcpx-nccl-plugin-installer",
+            image=(
+                "us-docker.pkg.dev/gce-ai-infra/gpudirect-tcpxo/nccl-plugin-gpudirecttcpx-dev:latest"
+            ),
+            command=command,
+            env=[{"name": "LD_LIBRARY_PATH", "value": "/usr/local/nvidia/lib64"}],
+            volumeMounts=volume_mounts,
+        )
+
+    def _build_a3_mega_tcpx_sidecar_container(self) -> Nested[Any]:
+        """Builds a sidecar container which is required by A3
+        for GPU to GPU RDMA like networking.
+
+        Returns:
+            A nested dict of the sidecar container.
+        """
+        volume_mounts = [
             {
                 "name": "nvidia-install-dir-host",
-                "hostPath": {"path": "/home/kubernetes/bin/nvidia/lib64"},
+                "mountPath": "/usr/local/nvidia/lib64",
             },
+            {
+                "name": "tcpx-socket",
+                "mountPath": "/run/tcpx",
+            },
+        ]
+
+        command = [
+            "bash",
+            "-c",
+            "set -ex; chmod 755 /fts/entrypoint_rxdm_container.sh; \n\
+            /fts/entrypoint_rxdm_container.sh --num_hops=2 --num_nics=8 \
+                --uid= --alsologtostderr &\n\
+            while [ ! -f /run/tcpx/terminated ]; do sleep 10; done;",
```

Comment:
for potential follow up, with k8s native sidecars, we should be able to remove this `while [ ! -f /run/tcpx/terminated ]; do sleep 10; done;"` and then launch the sidecar process in foreground.